### DISCUSSION TITLE
Glyphwise and Contextual Alternates

### DISCUSSION THREAD

**Discussion by A-D-C-O** [coldtype/coldtype#140] [Q&A]
I have a typeface that relies heavily on the contextual alternates (opentype "calt"). From what I can tell, coldtype seems to support these pretty well unless I use a Glyphwise structuring. I'm guessing that Glyphwise doesn't respect neighboring relationships since it treats a string as individual glyphs? Two questions here:

1. Am I doing something wrong? Is there a high-level way to ensure the use of calts when using Glyphwise, or...

2. Can I specify alts or even stylistic sets for specific characters in a string somehow?

Thanks!

**Comment by stenson** [comment]
Good question! I think it's kind of a grey area on how `Glyphwise` should work. Glyphwise builds up a text setting by setting neighboring triplets, which wouldn't respect complex chained calt relationships, since it's only ever looking at the relationship between the previous and the next character in a string (I think, I'm a little rusty on that).

For example, here's the same string set with a calt-dependent font with StSt, StSt (calt disabled), and Glyphwise:

<img width="945" alt="image" src="https://github.com/coldtype/coldtype/assets/89770/4df26210-0039-4989-a9f7-cd23cef85a8b">

The `Glyphwise` result (the blue outlines) really isn't correct in any way, since it doesn't build up the right chain of glyphs. In general, Glyphwise was built to preserve spacing/kerning for non-script variable fonts, so I wouldn't recommend its use in something like this, though it's certainly kind of interesting to create chaos in the opentype feature set with it. (If anyone reading this in the future has an idea of how to properly handle something like this, let me know!)

In general, setting alts or stylistic sets on neighboring letters in a string is tough, because every time you switch up those settings, you're really switching up the typeface completely, meaning you're interrupting the way opentype features are implemented across characters.

That doesn't mean you shouldn't do it, though, just that you'll need to do something different to get at it. So if you did want to tchange up features mid-word, you'll have to do something like this:

```python
from coldtype import *

@renderable((1080, 540), bg=1)
def scratch(r):
    return (P(
        StSt("Ca", "Cortado", fs:=500, features=dict(calt=False)),
        StSt("l", "Cortado", fs, ss03=1),
        StSt("ty", "Cortado", fs, ss04=1),
        )
        .distribute()
        .align(r, ty=1))
```

<img width="855" alt="image" src="https://github.com/coldtype/coldtype/assets/89770/cb023bd5-5687-46d2-82e6-76bee2e631c7">

Still pretty weird in a script context, since you'd probably want to maintain the calt relationships to get the designer's intent, but if you're working with a non-script typeface I could see this making a lot more sense.

(There's also a little bit of oddity here in that you'll need to set `features=dict(calt=0)` if you want to disable `calt`, as it isn't auto-recognized in the `kwargs` as a feature name, unlike `ss01`, `ss02` etc., which can either be specified as `ss01=1` or in a features dict, i.e. `features=dict(calt=0, ss01=1)`)
